scripts/gen_app_partitions.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+import sys
+import argparse
+import os
+import re
+import string
+from elf_helper import ElfHelper
+from elftools.elf.elffile import ELFFile
+
+
+# This script will create linker comands for power of two aligned MPU
+# when APP_SHARED_MEM is enabled.
+print_template = """
+/* Auto generated code do not modify */
+. = ALIGN( 1 << LOG2CEIL(data_smem_REPLACE_STRINGb_end - data_smem_REPLACE_STRING));
+data_smem_REPLACE_STRING = .;
+KEEP(*(SORT(data_smem_REPLACE_STRING*)))
+. = ALIGN(_app_data_align);
+data_smem_REPLACE_STRINGb_end = .;
+. = ALIGN( 1 << LOG2CEIL(data_smem_REPLACE_STRINGb_end - data_smem_REPLACE_STRING));
+"""
+
+full_list_of_partitions = []
+partitions_source_file= {}
+
+def find_partitions(filename):
+    with open(filename, 'rb') as f:
+        full_lib = ELFFile( f)
+        if (not full_lib):
+            print("Error parsing file: ",filename)
+            os.exit(1)
+
+        sections = [ x for x in full_lib.iter_sections()]
+        for section in sections:
+            if ("smem" in  section.name and not ".rel" in section.name):
+                partition_name = section.name.split("data_smem_")[1]
+                if partition_name not in full_list_of_partitions:
+                    full_list_of_partitions.append(partition_name)
+                    if args.verbose:
+                        partitions_source_file.update({partition_name: filename})
+
+def cleanup_remove_bss_regions():
+    for partition in full_list_of_partitions:
+        if (partition+"b" in full_list_of_partitions):
+            full_list_of_partitions.remove(partition+"b")
+
+def generate_final_linker(linker_file):
+    string= ''
+    for partition in full_list_of_partitions:
+            string += print_template.replace("REPLACE_STRING", partition)
+    with open(linker_file, "w") as fw:
+        fw.write(string)
+
+
+def parse_args():
+    global args
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument("-d", "--directory", required=True,
+                        help="Root build directory")
+    parser.add_argument("-o", "--output", required=False,
+                        help="Output ld file")
+    parser.add_argument("-v", "--verbose", action="count", default =0,
+                        help="Verbose Output")
+    args = parser.parse_args()
+
+
+def main():
+    parse_args()
+    root_directory = args.directory
+    linker_file = args.output
+    for dirpath, dirs, files in os.walk(root_directory):
+        for filename in files:
+            if re.match(".*\.obj$",filename):
```

Comment:
Pedantry: are we guaranteed that all generated ELF objects will be named "*.obj" on all platforms?  Seems likely enough, I guess, but would need Sebastien or someone else with more cmake knowledge to declare that for certain.

Reply:
I am assuming that the object files will be named as .obj. I believe we have the same assumption in other places inside our build.

Reply:
.obj is a reasonable assumption. We are already not portable to different suffixes. E.g. from zephyr/CMakeLists.txt

> set(OFFSETS_O_PATH ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/offsets.dir/arch/${ARCH}/core/offsets/offsets.c.obj)